src/sys/unix/mod.rs

```diff
@@ -5,6 +5,31 @@
 macro_rules! syscall {
     ($fn: ident ( $($arg: expr),* $(,)* ) ) => {{
         let res = unsafe { libc::$fn($($arg, )*) };
+        #[cfg(target_os = "hermit")]
+        if res < 0 {
+            let e = match -res {
+                libc::errno::EACCES => std::io::ErrorKind::PermissionDenied,
+                libc::errno::EADDRINUSE => std::io::ErrorKind::AddrInUse,
+                libc::errno::EADDRNOTAVAIL => std::io::ErrorKind::AddrNotAvailable,
+                libc::errno::EAGAIN => std::io::ErrorKind::WouldBlock,
+                libc::errno::ECONNABORTED => std::io::ErrorKind::ConnectionAborted,
+                libc::errno::ECONNREFUSED => std::io::ErrorKind::ConnectionRefused,
+                libc::errno::ECONNRESET => std::io::ErrorKind::ConnectionReset,
+                libc::errno::EEXIST => std::io::ErrorKind::AlreadyExists,
+                libc::errno::EINTR => std::io::ErrorKind::Interrupted,
+                libc::errno::EINVAL => std::io::ErrorKind::InvalidInput,
+                libc::errno::ENOENT => std::io::ErrorKind::NotFound,
+                libc::errno::ENOTCONN => std::io::ErrorKind::NotConnected,
+                libc::errno::EPERM => std::io::ErrorKind::PermissionDenied,
+                libc::errno::EPIPE => std::io::ErrorKind::BrokenPipe,
+                libc::errno::ETIMEDOUT => std::io::ErrorKind::TimedOut,
+                _ => panic!("Unknown error {}", res),
+            };
+            Err(std::io::Error::from(e))
+        } else {
+            Ok(res)
+        }
```

Comment:
Does `std::io::Error::last_os_error()` not work on HermitOS?

Reply:
Within `std` we determined the last error in a different way and ignore `std::io::Error::last_os_error()`. In the meantime we fixed it (see rust-lang/rust#121765). Consequently, the nightly compiler should support it. Should we use `last_os_error` or should we directly call our [abi](https://github.com/hermit-os/hermit-rs/blob/main/hermit-abi/src/lib.rs#L341-L342)?

Reply:
That's a though one. Is it common for Hermit to run on Rust stable? Or are most users on nightly? If the latter is true I think it's ok to be on nightly-only for now.

Reply:
The kernel use some nightly feature and the user-space is able to use stable. I think we can use the nightly compiler. I will fix it.